cfme/common/provider.py

```diff
@@ -1090,6 +1090,102 @@ def cleanup_vm(vm_name, provider):
+
+    The logic implied while waiting is:
+    1. Check if VM is in_desired_state, if so, we're done
+    2. If not, check if it is in_state_requiring_prep
+         (for example, you can't stop a suspended VM, so we need to prep the VM by starting it)
+    3. Move the VM to the correct 'prep state' (calling method do_prep)
+    4. Check if the VM is in a state that allows moving to the desired state
+         (using in_actionable_state check)
+    5. Perform the action to put the VM into that state (calling method do_action)
+
+    The methods are defined differently for each desired state.
+
+    For some states, step 2 and 3 do not apply, see for example when desired state is 'running',
+    in those cases just make sure that 'in_state_requiring_prep' always returns false.
+
+    Args:
+        provider: Provider class object
+        vm_name: Name of the VM/instance
+        state: str, one of "running"/"started", "stopped", "suspended", "paused"
+    """
+    vm = vm_name
+
+    if state in ["running", "started"]:
+        in_desired_state = lambda: provider.mgmt.is_vm_running(vm)
+        in_state_requiring_prep = lambda: False
+        in_actionable_state = (
+            lambda: provider.mgmt.is_vm_stopped(vm) or _is_vm_suspended_or_paused(provider, vm)
+        )
+        do_prep = lambda: None  # do_prep will never be needed for this state
+        do_action = lambda: provider.mgmt.start_vm(vm)
+
+    elif state == "stopped":
+        in_desired_state = lambda: provider.mgmt.is_vm_stopped(vm)
+        in_state_requiring_prep = lambda: _is_vm_suspended_or_paused(provider, vm)
+        in_actionable_state = lambda: provider.mgmt.is_vm_running(vm)
+        do_prep = lambda: provider.mgmt.start_vm(vm)
+        do_action = lambda: provider.mgmt.stop_vm(vm)
+
+    elif state == "suspended":
+        in_desired_state = lambda: provider.mgmt.is_vm_suspended(vm)
+        in_state_requiring_prep = (
+            lambda: provider.mgmt.is_vm_stopped(vm) or _is_vm_paused(provider, vm)
+        )
+        in_actionable_state = lambda: provider.mgmt.is_vm_running(vm)
+        do_prep = lambda: provider.mgmt.start_vm(vm)
+        do_action = lambda: provider.mgmt.suspend_vm(vm)
+
+    elif state == "paused":
+        in_desired_state = lambda: provider.mgmt.is_vm_paused(vm)
+        in_state_requiring_prep = (
+            lambda: provider.mgmt.is_vm_stopped(vm) or _is_vm_suspended(provider, vm)
+        )
+        in_actionable_state = lambda: provider.mgmt.is_vm_running(vm)
+        do_prep = lambda: provider.mgmt.start_vm(vm)
+        do_action = lambda: provider.mgmt.pause_vm(vm)
+
+    else:
+        raise ValueError("Invalid state '{}'".format(state))
+
+    def _wait():
+        if in_desired_state():
+            return True
+        elif in_state_requiring_prep():
+            do_prep()
+        elif in_actionable_state():
+            do_action()
+
+        logger.debug(
```

Comment:
Can be passed as into wait_for as a kwarg to be logged automatically: https://github.com/RedHatQE/wait_for/blob/master/wait_for/__init__.py#L99

Reply:
I'll pass in a message kwarg as well. But this debug log helps you keep track of the changing states as the wait_for keeps looping so I think it's helpful for it to stay. 'message' kwarg only seems to log a message at the start/end of wait_for?

Reply:
Hm, you're right - that message is only at start/finish of wait_for, I was thinking it was printed on each iteration of the wait loop. Good call!